Add failed tasks count to Mill's progress indicator

zelosleone opened the pull request:
* Add failed tasks count to Mill's progress indicator

This change introduces a mechanism to track and display the number of failed tasks during build execution. Key modifications include:

- Added `setFailedTasksCount` method to the `Logger` trait
- Updated `Execution` to count and report failed tasks
- Modified `PromptLogger` to include failed tasks count in the progress indicator
- Implemented `setFailedTasksCount` in `MultiLogger` and `PromptLogger`

fixes #4588

The new feature provides more visibility into build status by showing the number of failed tasks in the progress line, helping users quickly identify build issues.

Example output from the integration test:
```
[3346] public class Foo{
[3346]     public static String generateHtml(String text){
[3346]         Context context = new Context();
[3346]         context.setVariable("text", text);
[3346]         return new TemplateEngine().process("<h1 th:text=\"${text}\"></h1>", context);
[3346]     }
[3346]     public static void main(String[] args) {
[3346]         ArgumentParser parser = ArgumentParsers.newFor("template").build()
[3346]                 .defaultHelp(true)
[3346]                 .description("Inserts text into a HTML template");
[3346]
[3346]         parser.addArgument("-t", "--text")
[3346]                 .required(true)
[3346]                 .help("text to insert");
[3346]
[3346]         Namespace ns = null;
[3346]         try {
[3346]             ns = parser.parseArgs(args);
[3346]         }catch(Exception e){
[3346]             System.out.println(e.getMessage());
[3346]             System.exit(1);
[3346]         }
[3346]
[3346]         System.out.println(generateHtml(ns.getString("text")));
[3346]     }
[3346] }
[3346] '
[3346] [DEBUG] failedTasksCounter - modified content: 'package foo;
[3346]
[3346] import net.sourceforge.argparse4j.ArgumentParsers;
[3346] import net.sourceforge.argparse4j.inf.ArgumentParser;
[3346] import net.sourceforge.argparse4j.inf.ArgumentParserException;
[3346] import net.sourceforge.argparse4j.inf.Namespace;
[3346] import org.thymeleaf.TemplateEngine;
[3346] import org.thymeleaf.context.Context;
[3346]
[3346] public class Foo {{
[3346]     public static String generateHtml(String text){
[3346]         Context context = new Context();
[3346]         context.setVariable("text", text);
[3346]         return new TemplateEngine().process("<h1 th:text=\"${text}\"></h1>", context);
[3346]     }
[3346]     public static void main(String[] args) {
[3346]         ArgumentParser parser = ArgumentParsers.newFor("template").build()
[3346]                 .defaultHelp(true)
[3346]                 .description("Inserts text into a HTML template");
[3346]
[3346]         parser.addArgument("-t", "--text")
[3346]                 .required(true)
[3346]                 .help("text to insert");
[3346]
[3346]         Namespace ns = null;
[3346]         try {
[3346]             ns = parser.parseArgs(args);
[3346]         }catch(Exception e){
[3346]             System.out.println(e.getMessage());
[3346]             System.exit(1);
[3346]         }
[3346]
[3346]         System.out.println(generateHtml(ns.getString("text")));
[3346]     }
[3346] }
[3346] '
[3346] [DEBUG] failedTasksCounter - failed build isSuccess: false
[3346] [DEBUG] failedTasksCounter - failed build out: ''
[3346] [DEBUG] failedTasksCounter - failed build err: '============================== compile ==============================
[3346] [30/30] compile
[3346] [30] [info] compiling 1 Java source to C:\Users\deniz\Documents\GitHub\mill\out\integration\feature\full-run-logs\local\fork\test.dest\sandbox\run-4\out\compile.dest\classes ...
[3346] [30] [error] C:/Users/deniz/Documents/GitHub/mill/out/integration/feature/full-run-logs/local/fork/test.dest/sandbox/run-4/src/foo/Foo.java:11:5: illegal start of expression
[3346] [30/30, 1 failed] ============================== compile ============================== 4s
[3346] 1 tasks failed
[3346] compile javac returned non-zero exit code'
[3346] [DEBUG] failedTasksCounter - pattern matches: true
[3346] [DEBUG] failedTasksCounter - matches found: List([30/30, 1 failed])
```

zelosleone commented:
@lihaoyi solved the issue and provided an integration test just like you wanted, included the debug info just in case to not to take your time for nothing. 

For people who use windows and search through github to see how to run the integration tests:
$env:MILL_TEST_RESOURCE_DIR="$pwd\integration\feature\full-run-logs\resources"; $env:MILL_INTEGRATION_LAUNCHER="$pwd\mill.bat"; $env:MILL_INTEGRATION_SERVER_MODE="true"; .\mill.bat integration.feature[full-run-logs].local.fork.test

as one-liner command on powershell works.

lihaoyi commented:
Seems to partially work now! Still hitting some issues:

```scala
lihaoyi mill$ /Users/lihaoyi/Github/mill/mill-assembly.jar main.test.compile        
Mill version SNAPSHOT is different than configured for this directory!
Configured version is 0.12.8-1-46e216-native (/Users/lihaoyi/Github/mill/.config/mill-version)
[1002/1006] testkit.compile
[1002] [info] compiling 1 Scala source to /Users/lihaoyi/Github/mill/out/testkit/compile.dest/classes ...
[1002] [error] -- [E103] Syntax Error: /Users/lihaoyi/Github/mill/testkit/src/mill/testkit/UtestIntegrationTestSuite.scala:24:0 
[1002] [error] 24 |breakage
[1002] [error]    |^^^^^^^^
[1002] [error]    |Illegal start of toplevel definition
[1002] [error]    |
[1002] [error]    | longer explanation available when compiling with `-explain`
[1002] [error] one error found
[1006/1006] ============================== main.test.compile ==============================
1 tasks failed
testkit.compile Compilation failed
lihaoyi mill$ /Users/lihaoyi/Github/mill/mill-assembly.jar -k main.test.compile
Mill version SNAPSHOT is different than configured for this directory!
Configured version is 0.12.8-1-46e216-native (/Users/lihaoyi/Github/mill/.config/mill-version)
[1002/1006] testkit.compile
[1002] [info] compiling 1 Scala source to /Users/lihaoyi/Github/mill/out/testkit/compile.dest/classes ...
[1002] [error] -- [E103] Syntax Error: /Users/lihaoyi/Github/mill/testkit/src/mill/testkit/UtestIntegrationTestSuite.scala:24:0 
[1002] [error] 24 |breakage
[1002] [error]    |^^^^^^^^
[1002] [error]    |Illegal start of toplevel definition
[1002] [error]    |
[1002] [error]    | longer explanation available when compiling with `-explain`
[1002] [error] one error found
[1002] [1002] testkit.compile failed
[1004/1006] 
[1003/1006] 
[1004] [1004] main.test.upstreamCompileOutput failed
[1003] [1003] main.test.transitiveCompileClasspath failed
[1005/1006] 
[1005] [1005] main.test.compileClasspath failed
[1006/1006] 
[1006] [1006] main.test.compile failed
[1006/1006, 4 failed] ============================== main.test.compile ==============================
1 tasks failed
testkit.compile Compilation failed
```

- If I don't pass in `-k`, the `n failures` message does not properly show up
- When I do pass in `-k` it seems to count all the downstream tasks that were skipped as failures, even though only one task actually failed.

zelosleone commented:
> Seems to partially work now! Still hitting some issues:
> 
> ```scala
> lihaoyi mill$ /Users/lihaoyi/Github/mill/mill-assembly.jar main.test.compile        
> Mill version SNAPSHOT is different than configured for this directory!
> Configured version is 0.12.8-1-46e216-native (/Users/lihaoyi/Github/mill/.config/mill-version)
> [1002/1006] testkit.compile
> [1002] [info] compiling 1 Scala source to /Users/lihaoyi/Github/mill/out/testkit/compile.dest/classes ...
> [1002] [error] -- [E103] Syntax Error: /Users/lihaoyi/Github/mill/testkit/src/mill/testkit/UtestIntegrationTestSuite.scala:24:0 
> [1002] [error] 24 |breakage
> [1002] [error]    |^^^^^^^^
> [1002] [error]    |Illegal start of toplevel definition
> [1002] [error]    |
> [1002] [error]    | longer explanation available when compiling with `-explain`
> [1002] [error] one error found
> [1006/1006] ============================== main.test.compile ==============================
> 1 tasks failed
> testkit.compile Compilation failed
> lihaoyi mill$ /Users/lihaoyi/Github/mill/mill-assembly.jar -k main.test.compile
> Mill version SNAPSHOT is different than configured for this directory!
> Configured version is 0.12.8-1-46e216-native (/Users/lihaoyi/Github/mill/.config/mill-version)
> [1002/1006] testkit.compile
> [1002] [info] compiling 1 Scala source to /Users/lihaoyi/Github/mill/out/testkit/compile.dest/classes ...
> [1002] [error] -- [E103] Syntax Error: /Users/lihaoyi/Github/mill/testkit/src/mill/testkit/UtestIntegrationTestSuite.scala:24:0 
> [1002] [error] 24 |breakage
> [1002] [error]    |^^^^^^^^
> [1002] [error]    |Illegal start of toplevel definition
> [1002] [error]    |
> [1002] [error]    | longer explanation available when compiling with `-explain`
> [1002] [error] one error found
> [1002] [1002] testkit.compile failed
> [1004/1006] 
> [1003/1006] 
> [1004] [1004] main.test.upstreamCompileOutput failed
> [1003] [1003] main.test.transitiveCompileClasspath failed
> [1005/1006] 
> [1005] [1005] main.test.compileClasspath failed
> [1006/1006] 
> [1006] [1006] main.test.compile failed
> [1006/1006, 4 failed] ============================== main.test.compile ==============================
> 1 tasks failed
> testkit.compile Compilation failed
> ```
> 
>     * If I don't pass in `-k`, the `n failures` message does not properly show up
> 
>     * When I do pass in `-k` it seems to count all the downstream tasks that were skipped as failures, even though only one task actually failed.

Ups! That actually has an incredibly easy fix, i will post it with the updated integration test.

zelosleone commented:
@lihaoyi Should be fixed now, thanks for pointing it out. I forgot to test the skipped tasks.

lihaoyi commented:
I just tried the latest 631fa959bb817b797cf17074e44d3354323aa419, still not seeing the `n failures` message appear reliably when I break things locally and run `main.__.compile`. How are you testing this manually?

zelosleone commented:
> I just tried the latest [631fa95](https://github.com/com-lihaoyi/mill/commit/631fa959bb817b797cf17074e44d3354323aa419), still not seeing the `n failures` message appear reliably when I break things locally and run `main.__.compile`. How are you testing this manually?

Via the integration test and debug info i get, can you share your testing files so i can try with them instead?

lihaoyi commented:
Here's what I'm running to exercise this end-to-end

```bash
git clean -xdf
./mill dist.installLocal
echo breakage >> main/test/src/mill/TestMain.scala
ci/patch-mill-bootstrap.sh
./mill-assembly.jar main.__.test
```

The final command shows the failure in the logs but not in the prompt

```bash
lihaoyi mill$ ./mill-assembly.jar main.__.test
Mill version SNAPSHOT is different than configured for this directory!
Configured version is 0.12.8-1-46e216-native (/Users/lihaoyi/Github/mill/.config/mill-version)
[mill-build/build.mill-62/66] compile
[mill-build/build.mill-62] [info] compiling 1 Scala source to /Users/lihaoyi/Github/mill/out/mill-build/mill-build/compile.dest/classes ...
[mill-build/build.mill-62] [info] done compiling
[build.mill-62/66] compile
[build.mill-62] [info] compiling 29 Scala sources to /Users/lihaoyi/Github/mill/out/mill-build/compile.dest/classes ...
[build.mill-62] [warn] -- /Users/lihaoyi/Github/mill/website/package.mill:316:13
[build.mill-62] [warn] 316 │            millLastTag.split('.'),
[build.mill-62] [warn]     │            ^
[build.mill-62] [warn]     │method copyArrayToImmutableIndexedSeq in class LowPriorityImplicits2 is deprecated since 2.13.0: implicit conversions from Array to immutable.IndexedSeq are implemented by copying; use `toIndexedSeq` explicitly if you want to copy, or use the more efficient non-copying ArraySeq.unsafeWrapArray
[build.mill-62] [warn] one warning found
[build.mill-62] [info] done compiling
[2727/2985] main.test.compile
[2727] [info] compiling 4 Scala sources to /Users/lihaoyi/Github/mill/out/main/test/compile.dest/classes ...
[2727] [error] -- [E103] Syntax Error: /Users/lihaoyi/Github/mill/main/test/src/mill/TestMain.scala:6:0 
[2727] [error] 6 |breakage
[2727] [error]   |^^^^^^^^
[2727] [error]   |Illegal start of toplevel definition
[2727] [error]   |
[2727] [error]   | longer explanation available when compiling with `-explain`
[2727] [error] one error found
[2985/2985] ============================== main.__.test ============================== 26s
1 tasks failed
main.test.compile Compilation failed
```

zelosleone commented:
> Here's what I'm running to exercise this end-to-end
> 
> ```shell
> git clean -xdf
> ./mill dist.installLocal
> echo breakage >> main/test/src/mill/TestMain.scala
> ci/patch-mill-bootstrap.sh
> ./mill-assembly.jar main.__.test
> ```
> 
> The final command shows the failure in the logs but not in the prompt
> 
> ```shell
> lihaoyi mill$ ./mill-assembly.jar main.__.test
> Mill version SNAPSHOT is different than configured for this directory!
> Configured version is 0.12.8-1-46e216-native (/Users/lihaoyi/Github/mill/.config/mill-version)
> [mill-build/build.mill-62/66] compile
> [mill-build/build.mill-62] [info] compiling 1 Scala source to /Users/lihaoyi/Github/mill/out/mill-build/mill-build/compile.dest/classes ...
> [mill-build/build.mill-62] [info] done compiling
> [build.mill-62/66] compile
> [build.mill-62] [info] compiling 29 Scala sources to /Users/lihaoyi/Github/mill/out/mill-build/compile.dest/classes ...
> [build.mill-62] [warn] -- /Users/lihaoyi/Github/mill/website/package.mill:316:13
> [build.mill-62] [warn] 316 │            millLastTag.split('.'),
> [build.mill-62] [warn]     │            ^
> [build.mill-62] [warn]     │method copyArrayToImmutableIndexedSeq in class LowPriorityImplicits2 is deprecated since 2.13.0: implicit conversions from Array to immutable.IndexedSeq are implemented by copying; use `toIndexedSeq` explicitly if you want to copy, or use the more efficient non-copying ArraySeq.unsafeWrapArray
> [build.mill-62] [warn] one warning found
> [build.mill-62] [info] done compiling
> [2727/2985] main.test.compile
> [2727] [info] compiling 4 Scala sources to /Users/lihaoyi/Github/mill/out/main/test/compile.dest/classes ...
> [2727] [error] -- [E103] Syntax Error: /Users/lihaoyi/Github/mill/main/test/src/mill/TestMain.scala:6:0 
> [2727] [error] 6 |breakage
> [2727] [error]   |^^^^^^^^
> [2727] [error]   |Illegal start of toplevel definition
> [2727] [error]   |
> [2727] [error]   | longer explanation available when compiling with `-explain`
> [2727] [error] one error found
> [2985/2985] ============================== main.__.test ============================== 26s
> 1 tasks failed
> main.test.compile Compilation failed
> ```

Will check with your method, converted to draft to stop the running time of build to avoid wasting it, will ping you when i solve it

zelosleone commented:
@lihaoyi Finally finished debugging i think, the issue was multi-fold. 

- It was actually succesfully giving the info to headerprefix about failed tasks all the time, the problem was failed task count was getting lost during the background refresh while tasks were switching. Made sure its actually keeping task count. had to enable refreshprompt manually for a while for headerprefix on promptlogger to debug this.
- Made the task failure detection more strict, maybe more strict? Need your input on this as i also stopped uptime issues to be detected as failure. Also made sure we are not counting skipped, aborted tasks as failures. Added new test cases for this, specially useful while manually debugging.
- Made sure both interactive and CI modes works for failure tracking. both with --ticker and no ticker flags enabled.

Will be waiting for your input.

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -154,6 +155,29 @@ private[mill] case class Execution(
                   .toMap
 
+                // Check if this task has any upstream failures
+                val hasUpstreamFailures = upstreamResults.values.exists(r => !r.asSuccess.isDefined)
+
+                // Only count failures if there are no upstream failures (i.e. this is a root failure)
+                val currentFailedCount = if (!hasUpstreamFailures) {
+                  upstreamResults.values.count(r =>
+                    !r.asSuccess.isDefined &&
+                      !r.isInstanceOf[ExecResult.Skipped.type] &&
+                      !r.isInstanceOf[ExecResult.Aborted.type]
+                  )
+                } else 0
```
Could this be consolidated into something like

```scala
val hasUpstreamFailures = upstreamResults.values.exists(r => r.asFailing.isDefined)
```

zelosleone replied:
That would consider skipped and/or aborted tasks as failures as well

lihaoyi replied:
`Skipped` and `Aborted` don't inherit from `Failing`, so that should be ok

zelosleone replied:
Alright i will do it then

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -154,6 +155,29 @@ private[mill] case class Execution(
+                }
+                logger.setFailedTasksCount(rootFailedCount.get())
+
+                // Always include the failed count in header if there are failures
+                val headerPrefix = s"$countMsg$verboseKeySuffix${
+                    if (rootFailedCount.get() > 0) s", ${rootFailedCount.get()} failed" else ""
+                  }"
+                logger.setPromptHeaderPrefix(headerPrefix)
```
Do we need to run this logic twice, before and after each task is run? It seems to me that running it after each task should be sufficient?

zelosleone replied:
We definitely should, i tried it without running twice and failure count in header gets lost during the task count.

zelosleone replied:
This is solved!

lihaoyi commented on core/internal/src/mill/internal/PromptLogger.scala:
```diff
@@ -362,7 +371,8 @@ private[mill] object PromptLogger {
         now,
         startTimeMillis,
-        if (headerPrefix.isEmpty) "" else s"[$headerPrefix]",
+        if (headerPrefix.isEmpty) ""
+        else s"[$headerPrefix${if (failedTasksCount > 0) s", $failedTasksCount failed" else ""}]",
```
I'm seeing this `"$count failed"` string duplicated in a ew places; could we consolidate it

zelosleone replied:
Yeah we can do this, i will do it and send a commit.

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -15,42 +15,23 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
 
```
As mentioned before the test suite will need a cleanup so it's more understandable by future maintainers

zelosleone replied:
I will try to clean it up and include it in my commit.

lihaoyi commented:
Thanks @zelosleone! It looks pretty good, just needs some cleanup I think but don't see any major issues

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -188,6 +212,25 @@ private[mill] case class Execution(
+                if (newFailures > 0) {
+                  rootFailedCount.addAndGet(newFailures)
+                }
+                logger.setFailedTasksCount(rootFailedCount.get())
+
+                // Always show failed count in header if there are failures
+                logger.setPromptHeaderPrefix(s"$countMsg$verboseKeySuffix${
```
Also why do we need both `setFailedTasksCount` and `setPromptHeaderPrefix`? Seems to me that one or the other should be sufficient

zelosleone replied:
They serve different purposes to explain:
setFailedTasksCount: stores the raw count
setPromptHeaderPrefix: sets additional context information that appears before the failed count

lihaoyi replied:
Sure, but why should the logger need to store the failed count? Shouldn't it just need to know what prompt header prefix to display?

zelosleone replied:
There are several problems for this:

- The storage of numbers is actually needed since it adds up to the count with failures in total, and plays a vital part to actually show the task count accurately even with the prompt getting refreshed with tasks on background.

- We could also try to combine it but the above can prevent it from working, your suggestiong in the first place allows a cleaner, easier maintance of code so i sort of agree with your approach too but i didn't try to set it in the same header so i am not sure if it could work even with the reason i gave above. I would need to test it.

- Imo we could use task failure count by itself and call upon it on other things in the future maybe? This seperation could be beneficial in the future if we wanted something different as well.

I will go with your suggestion though, since i didn't maintain a large scala codebase like this before.

zelosleone replied:
@lihaoyi Will wait for your input on this to complete the cleanup.

lihaoyi replied:
I don't understand this:

> The storage of numbers is actually needed since it adds up to the count with failures in total, and plays a vital part to actually show the task count accurately even with the prompt getting refreshed with tasks on background.

Why does it need to know the failed task count to refresh? Isn't using the most recent `setPromptHeaderPrefix` enough?

zelosleone replied:
We are writing two times but i am not sure if its gonna work, that's my only concern because I didn't try it with your suggestion yet.

lihaoyi replied:
We shouldn't be redundantly writing the same data in random places until you stop seeing errors. You need to figure out why it wasn't working when written once and make it work

zelosleone replied:
Alright i will check it out.

zelosleone commented:
@lihaoyi Cleaned up the test and made a new util for failure count prefix on utils

lihaoyi commented on integration/feature/full-run-logs/resources/build.mill:
```diff
@@ -7,4 +7,12 @@ object `package` extends RootModule with JavaModule {
     ivy"org.slf4j:slf4j-simple:2.0.9"
   )
+
+  def testForkEnv = T {
+    Map(
+      "MILL_TEST_RESOURCE_DIR" -> resources().map(_.path).mkString(";"),
+      "MILL_EXECUTABLE_PATH" -> os.pwd.toString(),
+      "MILL_INTEGRATION_SERVER_MODE" -> "true"
+    )
+  }
```
We can't probably clean this up

zelosleone replied:
I will remove it then, I only used to run it via powershell on my laptop for easier access.

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -15,42 +15,23 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
-        .replace("<digits>", "\\E\\d+\\Q")
-        .replace("<dashes>", "\\E=+\\Q")
+      assert(res.out.contains("<h1>hello</h1>"))
 
```
Can these changes to existing tests be reverted? Those didn't have failures so nothing should change right?

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -74,5 +55,41 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
+    test("failedTasksCounter") - integrationTest { tester =>
+      import tester._
+
+      // First ensure clean compilation works
+      val cleanBuild = eval(("compile"))
+      cleanBuild.isSuccess ==> true
+
+      // Modify the Java source to introduce a compilation error
+      val javaSource = os.Path(workspacePath, os.pwd)
+      val fooJava = javaSource / "src" / "foo" / "Foo.java"
+      val originalContent = os.read(fooJava)
```
No need to check clean compile works, we have other tests for that. Can just start with broken code

zelosleone commented:
@lihaoyi Clean up is done, actually applying your suggestion made me remove one of the logic of always applying to headerprefix as well!

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -74,5 +74,27 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
+      val originalContent = os.read(fooJava)
+
+      // Introduce a compilation error by adding an unclosed brace
+      os.write.over(
+        fooJava,
+        data = originalContent.replace("class Foo", "class Foo {"),
+        createFolders = true
+      )
```
You can use the `modifyFile`  helper mere

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -74,5 +74,27 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
+
+      // Verify the output shows failed tasks count in the progress indicator
+      val expectedPattern = "\\[\\d+/\\d+,\\s*\\d+\\s*failed\\]".r // Matches [X/Y, N failed]
+      expectedPattern.findFirstIn(res.err).isDefined ==> true
```
Let's also capture the number of failed tasks and assert against it

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -154,6 +155,22 @@ private[mill] case class Execution(
+                val hasUpstreamFailures = upstreamResults.values.exists(r => r.asFailing.isDefined)
+
+                // Only count failures if there are no upstream failures (i.e. this is a root failure)
+                val currentFailedCount = if (!hasUpstreamFailures) {
+                  upstreamResults.values.count(r =>
+                    !r.asSuccess.isDefined &&
+                      !r.isInstanceOf[ExecResult.Skipped.type] &&
+                      !r.isInstanceOf[ExecResult.Aborted.type]
+                  )
+                } else 0
+
+                if (currentFailedCount > 0) {
+                  rootFailedCount.addAndGet(currentFailedCount)
+                }
```
This looks wrong. Why are we adding upstream failures to the  `rootFailedCount` when they should be added by the code snippet below when the upstream failure themselves just finished running?

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -74,5 +74,27 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
       assert(millProfile.exists(_.obj("label").str == "show"))
       assert(millChromeProfile.exists(_.obj("name").str == "show"))
     }
+    test("failedTasksCounter") - integrationTest { tester =>
```
Let's add an example with `--keep-going` to make sure we are correctly handling the `Skipped` case

lihaoyi commented on integration/feature/full-run-logs/src/FullRunLogsTests.scala:
```diff
@@ -74,5 +74,27 @@ object FullRunLogsTests extends UtestIntegrationTestSuite {
+      )
+
+      // Run with ticker to see the failed tasks count
+      val res = eval(("--ticker", "true", "compile"))
```
Let's run this with both `compile` and `jar` to make sure that both the selected task itself failing, as well as upstream tasks failing, correctly sets the fail count

zelosleone replied:
But we dont count upstream tasks failing as i mentioned before?

lihaoyi replied:
we should count upstream tasks failing, and AFAICT from the code your implementation does

zelosleone replied:
IMO my previous test was wrong, glad you made me work on it again

zelosleone commented:
@lihaoyi done with the changes as you have requested.

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -147,8 +148,14 @@ private[mill] case class Execution(
 
               val verboseKeySuffix = s"/${terminals0.size}"
               logger.setPromptHeaderPrefix(s"$countMsg$verboseKeySuffix")
```
It seems like all your problems with the `, failed` suffix disappearing is because of this line. If you update this line to contain the suffix, you won't need to put the weird code in the `failed.get()` branch to immediately overwrite it

lihaoyi commented on core/api/src/mill/api/Logger.scala:
```diff
@@ -86,4 +86,6 @@ trait Logger extends AutoCloseable {
   def withOutStream(outStream: PrintStream): Logger = this
   private[mill] def logPrefixKey: Seq[String] = Nil
+
+  private[mill] def setFailedTasksCount(count: Int): Unit = ()
```
I don't see this method and its various overrides getting called anywhere. Can we remove them?

zelosleone replied:
Oh yeah that was before the last commit, I will remove it.

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -188,7 +195,22 @@ private[mill] case class Execution(
 
-                if (failFast && res.newResults.values.exists(_.asSuccess.isEmpty))
+                // After execution, only count new failures if there were no upstream failures
+                val hasUpstreamFailures = upstreamResults.values.exists(r => r.asFailing.isDefined)
+                val newFailures = if (!hasUpstreamFailures) {
+                  res.newResults.values.count(r => r.asFailing.isDefined)
+                } else 0
```
This can be simplified

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -188,7 +195,22 @@ private[mill] case class Execution(
 
-                if (failFast && res.newResults.values.exists(_.asSuccess.isEmpty))
+                // After execution, only count new failures if there were no upstream failures
+                val hasUpstreamFailures = upstreamResults.values.exists(r => r.asFailing.isDefined)
```
Do we need this branch? If there are upstream failures, the current task should never return a failure, and should be `Skipped`

lihaoyi commented on core/internal/src/mill/internal/PromptLogger.scala:
```diff
@@ -334,13 +338,22 @@ private[mill] object PromptLogger {
+
+    def getHeaderPrefix() = headerPrefix
+
+    def setFailedTasksCount(count: Int): Unit = {
+      failedTasksCount = count
+    }
```
This can also be cleaned up, along with the backing field?

zelosleone replied:
Nope, they are being used as formatting. Internal references are there to be used with display.

zelosleone replied:
Specially that's the only way we could update header with error count with failfast.

lihaoyi replied:
I'm running `Cmd-F` `setFailedTasksCount` in my chrome browser on this PR and I don't see it called anywhere

zelosleone replied:
Huh, i tried to set it for setting the formatting for display, i will check.

zelosleone replied:
Yeah specially that part was not needed, but only that part. I will remove it and set the private var failedtaskmethod to val instead. Other internal methods are needed to set displaying correctly.

lihaoyi commented on core/internal/src/mill/internal/PromptLogger.scala:
```diff
@@ -334,13 +338,18 @@ private[mill] object PromptLogger {
       .empty[Seq[String], Status](PromptLoggerUtil.seqStringOrdering)
 
     private var headerPrefix = ""
+    private val failedTasksCount = 0
```
Isn't this always zero now?

zelosleone replied:
Yeah its always zero, in there. Its passed into setPromptHeaderPrefix eventually and overall of them i left behind are needed for display formatting.

lihaoyi replied:
If it's always zero, lets just get rid of it and inline 0 at the callsites. That should let us simplify the code further

zelosleone replied:
Alright done that but ngl Li, i didnt think about that, i learnt a great deal of functional programming (not just scala) here, thanks.

lihaoyi replied:
No problem! It's expected there'll be some friction entering a new codebase, that's all just part of the process

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -146,9 +147,14 @@ private[mill] case class Execution(
+                  mill.internal.Util.formatFailedCount(rootFailedCount.get())
+                }")
+              if (failed.get()) {
+                // Even when we're in fail-fast mode and a previous task has failed,
```
This comment is stale

zelosleone commented:
@lihaoyi All CI Tests also pass. Updated the last comment  as well.

zelosleone commented:
@lihaoyi We should re-run the failed task, it seems like a github issue as it happened to me on my own repo before and now as well. It passes after re-running the test, it happened with the same error message before as well.

lihaoyi commented on core/internal/src/mill/internal/PromptLogger.scala:
```diff
@@ -362,7 +364,8 @@ private[mill] object PromptLogger {
         startTimeMillis,
-        if (headerPrefix.isEmpty) "" else s"[$headerPrefix]",
+        if (headerPrefix.isEmpty) ""
+        else s"[$headerPrefix${Util.formatFailedCount(0)}]",
```
This `${Util.formatFailedCount(0)}` can be removed

zelosleone replied:
Nice, solved this as well

zelosleone commented:
Alright, for the next batch of reviews i will wait for you to finish all of them to start changing it. Otherwise i will take too much build-time i think. @lihaoyi

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -188,7 +192,19 @@ private[mill] case class Execution(
+                // Count new failures - if there are upstream failures, tasks should be skipped, not failed
+                val newFailures = res.newResults.values.count(r => r.asFailing.isDefined)
+
+                if (newFailures > 0) {
```
This conditional isn't necessary right? If it's zero then `addAndGet` is a no-op already, and it will never be negative

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -146,9 +147,12 @@ private[mill] case class Execution(
-              logger.setPromptHeaderPrefix(s"$countMsg$verboseKeySuffix")
-              if (failed.get()) None
-              else {
+              logger.setPromptHeaderPrefix(s"$countMsg$verboseKeySuffix${
+                  mill.internal.Util.formatFailedCount(rootFailedCount.get())
+                }")
```
Let's move this big string into a local `def` that we use in these two places to remove the duplication

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -146,9 +147,12 @@ private[mill] case class Execution(
+              logger.setPromptHeaderPrefix(s"$countMsg$verboseKeySuffix${
+                  mill.internal.Util.formatFailedCount(rootFailedCount.get())
+                }")
+              if (failed.get()) {
```
revert formatting change

lihaoyi replied:
This still isn't done

zelosleone replied:
this does not exist anymore, it was autofix bot's commit and i already changed that line several times for this to exist

lihaoyi commented on core/internal/src/mill/internal/PromptLogger.scala:
```diff
@@ -362,7 +364,8 @@ private[mill] object PromptLogger {
         now,
         startTimeMillis,
-        if (headerPrefix.isEmpty) "" else s"[$headerPrefix]",
+        if (headerPrefix.isEmpty) ""
```
revert formatting change

lihaoyi commented:
Some last nits, after that just need to do some manual testing on my end and then we can merge this

zelosleone commented:
> Some last nits, after that just need to do some manual testing on my end and then we can merge this

Alright done with these as well, will be waiting for the manual testing results.

lihaoyi commented on core/exec/src/mill/exec/Execution.scala:
```diff
@@ -188,7 +193,15 @@ private[mill] case class Execution(
+                // Always show failed count in header if there are failures
+                logger.setPromptHeaderPrefix(formatHeaderPrefix(countMsg, verboseKeySuffix))
+
+                if (failFast && res.newResults.values.exists(r => r.asFailing.isDefined))
```
This line should not need to be changed, it is unrelated to printing the failed suffix in the prompt logger

zelosleone replied:
failfast updates the header if we go without --keep going, that should stay, logger one's deletation will be impossible to recover, they are both relevant and deleting them will cause entire ticker system going down, i just tried with your recommendition and that did happen so yeah.

lihaoyi replied:
I'm no referring to removing the entire block, just reverting this one line change to the `if` condition that controls `failed.set(true)`. 

This is conditional controls the fail-fast behavior regardless of prompt or not and the previous behavior was correct already. The new behavior is fine too, but since it's unrelated to this PR we should clean it up

zelosleone replied:
Alright should be fine now

lihaoyi commented on core/internal/src/mill/internal/PromptLogger.scala:
```diff
@@ -341,6 +341,8 @@ private[mill] object PromptLogger {
 
     def getCurrentPrompt() = currentPromptBytes
 
+    def getHeaderPrefix() = headerPrefix
```
This is unused

lihaoyi commented on core/internal/src/mill/internal/Util.scala:
```diff
@@ -4,4 +4,12 @@ private[mill] object Util {
     char.toString * (targetLength - s.length) + s
   }
+
+  /**
+   * Format a failed count as a string to be used in status messages.
+   * Returns ", N failed" if count > 0, otherwise an empty string.
+   */
+  def formatFailedCount(count: Int): String = {
+    if (count > 0) s", $count failed" else ""
+  }
 }
```
Let's move this to the `Execution` companion object since that's where it's used

lihaoyi commented:
Manual testing seems to work, left more comments for cleanup

zelosleone commented:
@lihaoyi left 2 comments unresolved, please explain them further as i don't see the formatting needs in my git commit history.

lihaoyi commented:
those two lines are changes on this PR and should not be

zelosleone commented:
@lihaoyi First one doesn't exist anymore, i saw now it was the bot's automatic linting fix, don't need to revert the commit to 8 commits behind just for that, i changed these lines already.

For the second one, i explained my reasoning and i think there is a misunderstanding/miscommunication happening, best thing to do would be to suggest a modification for me to commit directly, i think github allows this, then i can try your suggestion localy and if there are errors, either try to solve them and/or tell you.

lihaoyi commented:
Thank @zelosleone! please email me your international bank transfer details and I will send you the bounty

zelosleone commented:
> Thank @zelosleone! please email me your international bank transfer details and I will send you the bounty

Thank you! Just emailed.
